### PR TITLE
fix(@aws-amplify/interactions): fix interactions v3 bugs and refactor type

### PR DESCRIPTION
**Pull request by wlee221**
_Issue #, if available:_ #6324

_Description of changes:_ This PR fixes current bugs with the Interactions category and strongly types the `Interactions.send()` request.

#### Bug fixes:
- Convert `aws-sdk-js-v3` response from `Lex.PostContent` to `Uint8Array`.
- Fix ReferenceError in react native chatbot ui. 
- Resume `AudioContext` when user presses the voice chat button on browser ui components.

#### Unit Test:
- Fix a bug where `jest.runAllTimers()` were called prematurely. This prevented `onComplete` callbacks from running and jest passed the tests vacuously.
- Fix a bug where jest was falsely expecting `onComplete` callback from text messages for voice messages.
- Test `Uint8Array` conversion.

#### Category Refactor:
- Strongly type `Interactions.send` API request type.
- Overload `Interactions.send`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/interactions/src/types/Interactions.ts [hunk 1]
```diff
@@ -14,9 +14,20 @@ export interface InteractionsOptions {
...
+	content: string;
 	options: {
-		[key: string]: string;
+		messageType: 'text';
 	};
-}
+};
+
+export type InteractionsVoiceMessage = {
+	content: string | object;
```
Comment (elorzafe) [inline packages/interactions/src/types/Interactions.ts:25]: Should this be only object?

File: packages/interactions/src/Providers/AWSLexProvider.ts [hunk 2]
```diff
@@ -23,6 +27,7 @@ import {
 	Credentials,
 	getAmplifyUserAgent,
 } from '@aws-amplify/core';
+import { convert } from './AWSLexHelper/convert';
```
Comment (elorzafe) [inline packages/interactions/src/Providers/AWSLexProvider.ts:30]: Maybe I will call it `AWSLexProviderHelper`

File: packages/interactions/src/Providers/AWSLexHelper/convert.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	}
+};
+
+const readOnBrowser = (stream: ReadableStream | Blob) => {
+	return new Response(stream)
+		.arrayBuffer()
+		.then(buffer => new Uint8Array(buffer));
+};
+
+// Node.js support
```
Comment (elorzafe) [inline packages/interactions/src/Providers/AWSLexHelper/convert.ts:18]: I will remove that line, Amplify doesn't officially support Node.js. Method name is clear enough. In which scenarios do you think that will be used?

File: packages/interactions/__tests__/Interactions-unit-test.ts [hunk 4]
```diff
@@ -556,46 +605,77 @@ describe('Interactions', () => {
...
+						m1: 'hi',
+						m2: 'done',
 					},
+					audioStream: typedStream,
 				});
+				jest.runAllTimers();
+			});
 
-				const textResponse = await interactions.send(
+			test('Interactions configuration and send voice message to existing bot and call onComplete from configure onComplete', async () => {
```
Comment (elorzafe) [inline packages/interactions/__tests__/Interactions-unit-test.ts:631]: I would rename the test to something more concise (what is testing, rather what is doing). Like  `onComplete callback from configure being called`

File: packages/interactions/__tests__/Interactions-unit-test.ts [hunk 5]
```diff
@@ -475,57 +495,88 @@ describe('Interactions', () => {
...
+						m1: 'voice:hi',
+						m2: 'voice:done',
 					},
+					audioStream: typedStream,
 				});
+				jest.runAllTimers();
 			});
 
-			test('Interactions configuration and send message to existing bot and call onComplete from configure onComplete', async () => {
+			test('Interactions configuration and send text message to existing bot and call onComplete from configure onComplete', async () => {
```
Comment (elorzafe) [inline packages/interactions/__tests__/Interactions-unit-test.ts:572]: I would rename it as my other comment

File: packages/interactions/__tests__/Interactions-unit-test.ts [hunk 6]
```diff
@@ -433,11 +452,13 @@ describe('Interactions', () => {
...
 				dialogState: 'ElicitSlot',
 				message: 'echo:hi',
+				audioStream: typedStream,
 			});
 		});
 
 		describe('Sending messages to bot', () => {
-			test('Interactions configuration and send message to existing bot and call onComplete from Interaction.onComplete', async () => {
+			jest.useFakeTimers();
+			test('Interactions configuration and send text message to existing bot and call onComplete from Interaction.onComplete', async () => {
```
Comment (elorzafe) [inline packages/interactions/__tests__/Interactions-unit-test.ts:461]: I would rename it to something shorter (check my other comment)

File: packages/interactions/src/types/Interactions.ts (hunk 1 above)
Comment (wlee221) [reply packages/interactions/src/types/Interactions.ts:25]: I thought so, but content (to Lex.PostContent) can be `string | object` as specified [here](https://github.com/aws/aws-sdk-js-v3/blob/master/clients/client-lex-runtime-service/commands/PostContentCommand.ts#L25).

File: packages/interactions/src/Providers/AWSLexProvider.ts (hunk 2 above)
Comment (wlee221) [reply packages/interactions/src/Providers/AWSLexProvider.ts:30]: Sounds good!

File: packages/interactions/__tests__/Interactions-unit-test.ts (hunk 4 above)
Comment (wlee221) [reply packages/interactions/__tests__/Interactions-unit-test.ts:631]: Sure!

File: packages/interactions/src/Providers/AWSLexHelper/convert.ts (hunk 3 above)
Comment (wlee221) [reply packages/interactions/src/Providers/AWSLexHelper/convert.ts:18]: That makes sense; I just added this for basic Node.js coverage (also helps jest) but not something to be actively used / maintained. I'll remove the line as suggested.

File: packages/interactions/src/types/Interactions.ts (hunk 1 above)
Comment (wlee221) [reply packages/interactions/src/types/Interactions.ts:25]: After looking at the [API](https://docs.aws.amazon.com/lex/latest/dg/API_runtime_PostContent.html), it seems that it is safe to say content would be only `object`. The only problem is that our unit tests send `string` content when sending voice messages. Let me see if I can fix that.

Comment (wlee221) [reply packages/interactions/src/types/Interactions.ts:25]: Unit tests are running fine, so I'll change as requested :D

**Comment by wlee221** [comment]
@jordanranz, @elorzafe thanks for the review!

- I will write a shorter typescript version of `aws-lex-audio` and consolidate into Interactions. That way, there will be only one `aws-lex-audio` in our libraries. There's also [content security concern](https://github.com/aws-amplify/amplify-js/issues/5024) with the file, so I'll work on that too.
- This has not been tested with Next.js, but I can test it (good time to learn Next.js also) after this merge.

**Pull request merged**